src/hyperlight_host/src/hypervisor/mod.rs

```diff
@@ -329,6 +318,66 @@ impl VirtualCPU {
     }
 }
 
+/// A trait for handling interrupts to a sandbox's vcpu
+pub trait InterruptHandle: Send + Sync {
+    /// Interrupt the corresponding sandbox from running.
+    ///
+    /// - If this is called while the vcpu is running, then it will interrupt the vcpu and return `true`.
+    /// - If this is called while the vcpu is not running, (for example during a host call), the
+    ///     vcpu will not immediately be interrupted, but will prevent the vcpu from running **the next time**
+    ///     it's scheduled, and returns `false`.
+    ///
+    /// # Note
+    /// This function will block for the duration of the time it takes for the vcpu thread to be interrupted.
+    fn kill(&self) -> bool;
+
+    /// Returns true iff the corresponding sandbox has been dropped
+    fn dropped(&self) -> bool;
+}
+
+#[cfg(any(kvm, mshv))]
+#[derive(Debug)]
+pub(super) struct LinuxInterruptHandle {
+    /// Invariant: vcpu is running => `running` is true. (Neither converse nor inverse is true)
+    running: AtomicBool,
+    /// Invariant: vcpu is running => `tid` is the thread on which it is running.
+    tid: AtomicU64,
+    /// True when an "interruptor" has requested the VM to be cancelled. Set immediately when
+    /// `kill()` is called, and cleared when the vcpu is no longer running.
+    /// This is used to
+    /// 1. make sure stale signals do not interrupt the
+    ///     the wrong vcpu (a vcpu may only be interrupted iff `cancel_requested` is true),
+    /// 2. ensure that if a vm is killed while a host call is running,
+    ///     the vm will not re-enter the guest after the host call returns.
+    cancel_requested: AtomicBool,
+    /// Whether the corresponding vm is dropped
+    dropped: AtomicBool,
+}
+
+#[cfg(any(kvm, mshv))]
+impl InterruptHandle for LinuxInterruptHandle {
+    fn kill(&self) -> bool {
+        self.cancel_requested.store(true, Ordering::Relaxed);
+
+        let signal_number = libc::SIGRTMIN() + INTERRUPT_VCPU_SIGRTMIN_OFFSET;
+        let mut sent_signal = false;
+
+        while self.running.load(Ordering::Relaxed) {
+            log::info!("Sending signal to kill vcpu thread...");
+            sent_signal = true;
+            unsafe {
+                libc::pthread_kill(self.tid.load(Ordering::Relaxed) as _, signal_number);
+            }
+            std::thread::sleep(std::time::Duration::from_micros(50));
```

Comment:
Before, this was 500 micros: https://github.com/hyperlight-dev/hyperlight/blob/a68ff5eac911cb6c89ecb1aade5f810018e57ab9/src/hyperlight_host/src/hypervisor/hypervisor_handler.rs#L766 . Should we make this configurable?

Reply:
Good idea. Updated in latest commit